sdk/python/lib/pulumi/provider/experimental/property_value.py

```diff
@@ -0,0 +1,499 @@
+        """
+        Creates a null property value.
+        """
+        return PropertyValue(None)
+
+    def with_secret(self, is_secret: bool) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified secret status.
+        """
+        return PropertyValue(
+            self.value,
+            is_secret=is_secret,
+            dependencies=self.dependencies,
+        )
+
+    def with_dependencies(self, dependencies: Iterable[str]) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified dependencies.
+        """
+        return PropertyValue(
+            self.value,
+            is_secret=self.is_secret,
+            dependencies=dependencies,
+        )
+
+    def with_value(self, value: PythonValue) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified value.
+        """
+        return PropertyValue(
+            value,
+            is_secret=self.is_secret,
+            dependencies=self.dependencies,
+        )
+
+    @property
+    def type(self) -> PropertyValueType:
+        """
+        Determines the type of the property value.
+        """
+        if self.value is None:
+            return PropertyValueType.NULL
+        if isinstance(self.value, bool):
+            return PropertyValueType.BOOL
+        if isinstance(self.value, (int, float)):
+            return PropertyValueType.NUMBER
+        if isinstance(self.value, str):
+            return PropertyValueType.STRING
+        if isinstance(self.value, pulumi.Asset):
+            return PropertyValueType.ASSET
+        if isinstance(self.value, pulumi.Archive):
+            return PropertyValueType.ARCHIVE
+        if isinstance(self.value, list):
+            return PropertyValueType.ARRAY
+        if isinstance(self.value, dict):
+            return PropertyValueType.OBJECT
+        if isinstance(self.value, ResourceReference):
+            return PropertyValueType.RESOURCE
+        if isinstance(self.value, Computed):
+            return PropertyValueType.COMPUTED
+        raise ValueError(f"Unsupported value type: {type(self.value)}")
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, PropertyValue):
+            return False
+        return (
+            self.value == other.value
+            and self.is_secret == other.is_secret
+            and self.dependencies == other.dependencies
+        )
+
+    def __hash__(self) -> int:
+        return hash([self.value, self.is_secret, self.dependencies])
+
+    def __str__(self) -> str:
+        v = str(self.value)
+        # If we have secretness or dependencies, we want to include that in the string representation.
+        if self.is_secret:
+            v = f"{v} (secret)"
+        if self.dependencies:
+            deps = ", ".join(self.dependencies)
+            v = f"{v} (dependencies: {deps})"
+        return v
+
+    def all_dependencies(self) -> Set[str]:
```

Comment:
cc @iwahbe 
Have you not needed a "collect all nested dependencies" function in the Go SDK yet? Is AllDependencies a sensible name to use for here and that?

Reply:
We havn't yet, but I'm sure we will eventually.